### ISSUE TITLE
A card token must start with tkn_. Field: cardToken (API)

### ISSUE THREAD

**Issue by BergMedia** [mollie/magento2#360]
It is impossible to place an order when the user creates a new account or uses a customer address from the address book.
The card token is empty. We use now the 1.22.0, but we have investigated the failure with 1.19 / 1.20 / 1.21 – all tests under the sandbox account. 

- Magento Version 2.3.6-p1 / Open source
- Mollie version 1.22.0:
- Latest OneStepCheckout 1.2.,44

`Mollie.ERROR: error: [2021-02-23T16:34:55+0000] Error executing API call (422: Unprocessable Entity): A card token must start with tkn_. Field: cardToken. Documentation: https://docs.mollie.com/reference/v2/payments-api/create-payment [] []`

**Comment by Frank-Magmodules** [comment]
HI @BergMedia ,

Thank you for opening and reporting this issue. As we can't reproduce this in our environment, can you please share your store details through our [contact form](https://www.magmodules.nl/support-form.html) so we can take a look with you on this and help you to solve this issue?

Thanks!

**Comment by wezzy** [comment]
Same problem here with PayPal. I've checked the mollie log and I've seen that sometimes when the user choose PayPal as a payment (we use PayPal through mollie not directly) the extension send a request with this key in the JSON:

`  "payment": {
    "cardToken": ""
  }`

and the API replies with: 

Mollie.ERROR: error: Error executing API call (422: Unprocessable Entity): A card token must start with tkn_. Field: payment.cardToken. Documentation: https://docs.mollie.com/reference/v2/payments-api/create-payment [] []

Searching in the same log, most of the times the PayPal request works and when it works the extension doesn't send any "payment" attribute in the JSON parameters.

**Comment by Frank-Magmodules** [comment]
Hi @wezzy ,

Also thank you for commenting on this issue. As we still can't reproduce this in a vanilla Magento 2 environment, please share your store details through our [contact form](https://www.magmodules.nl/support-form.html) so we can take a look with you to debug and fix this issue.

Thanks in forward!

**Comment by BergMedia** [comment]
Hi Frank,

I have modified your module with a simple IF statement after I realized where the problem is. On every request (load checkout page), it set the payment method to „creditcard“ and the FKN_. Parameter, because this is our Standard payment.
But when a user changes the payment method (around 75% of Sale), the extension dosent removes the TKN_. field from the „addtional_information“ array in the payment method.

I have modified the module with a simple IF statement after I realized where the problem is. On every request (load checkout page), it set the payment method to „creditcard“ and the FKN_. Parameter, because this is our Standard payment.
But when a user changes the payment method (around 75% of Sale), the extension dosent removes the TKN_. field from the „addtional_information“ array in the payment method.

I have made no Plugin for that – but set the composer instance of your module to lock at 1.0.22 :
 
Source/vendor/mollie/magento2/Controller/Checkout/Redirect.php

Line 124:
 
```
            $method = $order->getPayment()->getMethod();
            if($method!=='mollie_methods_creditcard') {
                $order->getPayment()->setAdditionalInformation(array());
            }
 
            $methodInstance = $this->paymentHelper->getMethodInstance($method);  
```
 
With this changes the extension can be compatible with the onestepCheckout (Norevegen - onestepcheckout.com).

Can you confirm that?

**Comment by BergMedia** [comment]
Hi @wezzy,

take a look in the rewrite.php and trace the $order->getPayment()->getAdditionalInformation(). When the array has the `tkn_. Field` - unset it.

**Comment by wezzy** [comment]
> Hi @wezzy,
> 
> take a look in the rewrite.php and trace the $order->getPayment()->getAdditionalInformation(). When the array has the `tkn_. Field` - unset it.

Thanks for your suggestion, I'll try ASAP.

**Comment by Frank-Magmodules** [comment]
Hi @BergMedia, @wezzy,
We are happy to share that we've just released the new [1.23.1](https://github.com/mollie/magento2/releases/tag/v1.23.1) version with a fix for this issue. 
Thank you again for the detailed issue report and for sharing the environment! 

We are closing this issue now but please feel free to reopen the issue if this still occurs.
